### PR TITLE
Move CI to GitHub Actions

### LINKED ISSUE #156: circleci tests queuing
The circleci tests for any given push are still constantly failing because the tests are being queued (not running) with the notice, "This job was waitlisted due to your concurrency limits". The tests are delayed from starting by 5 - 9 minutes, then time out at ~10 minutes and so 'fail'. You have to re-run tests multiple times to get all of them to pass, which can take 30 minutes - > 1 hr for a single push.

The circleci account plan says we can have "1 concurrent job with access to Linux and Windows builds on Small and Medium sizes". 

I googled and don't know why this queuing started happening seemingly randomly. I found this -  https://support.circleci.com/hc/en-us/articles/360055741771-Build-Not-Running-due-to-concurrency-limit-but-no-other-job-is-running - but I don't see any background jobs that are running which would cause the tests to queue.

Unless someone sees a quick fix, we should open a support ticket on circleci.

### LINKED ISSUE #157: Move to GitHub Actions
Move CI pipeline to GitHub Actions.

Follow what I did in `uplot` (See PR https://github.com/mtazzari/uvplot/pull/45) and issue https://github.com/mtazzari/uvplot/issues/44  to migrate from Travis CI.

### PR DESCRIPTION
**Pull request by mtazzari**
Moves CI from CircleCI to GitHub Actions to Fix #157, Fix #156.

Requirements:
- [x] run tests on GH Actions
- [x] build docs on GH Actions
- [x] deploy docs on GH Actions
- ~~(possibly) deploy to PyPI on master branch and tagged commits.~~ (slight time saver but not critical; could always do this in future PR if needed -JJ)

Using this guide to automate push to TestPyPI and PyPI:
https://packaging.python.org/guides/publishing-package-distribution-releases-using-github-actions-ci-cd-workflows/

### CODE REVIEW

**Comment by mtazzari** [comment]
@jeffjennings could you please add me as a mantainer for frank on https://test.pypi.org/manage/projects/
I need that to test uploading to PyPI.
Thanks.

**Comment by mtazzari** [comment]
It makes sense to have a `build-docs` job and a `deploy-docs` job in the same `Docs` workflow. 
In order to pass artefacts (eg built docs) between jobs: https://docs.github.com/en/actions/guides/storing-workflow-data-as-artifacts#passing-data-between-jobs-in-a-workflow

Next step is implement this.

**Comment by jeffjennings** [comment]
> @jeffjennings could you please add me as a mantainer for frank on https://test.pypi.org/manage/projects/ ?
> I need that to test uploading to PyPI.
> Thanks.

Just invited you. On test.pypi.org, the package name is frank-astro (because frank was taken by another person), so just change the package name in setup.py.

**Comment by mtazzari** [comment]
Great, thank you @jeffjennings 
I see, I'll alter the package name on-the-fly to test upload to TestPyPi.

**Comment by jeffjennings** [comment]
Yeah sorry it's annoying; I think you could also just change the .tar.gz name in `dist/`

**Comment by jeffjennings** [comment]
Hey @mtazzari, is there still more you want to do with this PR?

File: .github/workflows/docs.yml [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+  build:
+    name: Build docs
+    # Only proceed if the closed PR was also merged
+    # (see https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#running-your-workflow-when-a-pull-request-merges)
+    if_merged:
+        if: github.event.pull_request.merged == true
+    runs-on: ubuntu-latest
+
+    steps:
+      # Checks out a copy of your repository on the ubuntu-latest machine
```
Comment (rbooth200) [inline .github/workflows/docs.yml:21]: From [here](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#running-your-workflow-when-a-pull-request-merges) I thought that everything to be run needed to be under the if merged statement. What do you think?

**Comment by rbooth200** [comment]
I'm trying to get up to speed here, but I had a few queries before I can give this the OK. 

1. Should we run require that the docs build passes before we allow a branch to be merged? I'd think that we should check that we can build the docs each time at least.
2. I couldn't work out if/how we are deploying the docs to github.io. Is that working?

File: .github/workflows/docs.yml [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    types: [closed]
+
+jobs:
+  build:
+    name: Build docs
+    # Only proceed if the closed PR was also merged
+    # (see https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#running-your-workflow-when-a-pull-request-merges)
+    if_merged:
+        if: github.event.pull_request.merged == true
+    runs-on: ubuntu-latest
```
Comment (rbooth200) [inline .github/workflows/docs.yml:18]: Are you sure this is correct? I thought from [here](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#running-your-workflow-when-a-pull-request-merges) that the steps etc needed to be under the if merged. Maybe I'm wrong though..

Comment (jeffjennings) [reply .github/workflows/docs.yml]: Yes you're right, thanks for catching that. Fixed in [971c2f9]

**Comment by jeffjennings** [comment]
> I'm trying to get up to speed here, but I had a few queries before I can give this the OK.
> 
> 1. Should we run require that the docs build passes before we allow a branch to be merged? I'd think that we should check that we can build the docs each time at least.

Yeah, ok, I'm setting up this check each time a review is requested. 

> 2. I couldn't work out if/how we are deploying the docs to github.io. Is that working?

I thought so, but now see it's not yet. I'll try to get that to work and follow up here.

**Review by jeffjennings — COMMENTED** [review]
<test review>

**Review by jeffjennings — COMMENTED** [review]
<test>

**Review by jeffjennings — COMMENTED** [review]
<test>

**Review by jeffjennings — COMMENTED** [review]
<test>

**Review by jeffjennings — COMMENTED** [review]
<test>

**Review by jeffjennings — COMMENTED** [review]
<test>

**Comment by jeffjennings** [comment]
> > I'm trying to get up to speed here, but I had a few queries before I can give this the OK.
> > 
> > 1. Should we run require that the docs build passes before we allow a branch to be merged? I'd think that we should check that we can build the docs each time at least.
> 
> Yeah, ok, I'm setting up this check each time a review is requested.
> 
> > 2. I couldn't work out if/how we are deploying the docs to github.io. Is that working?
> 
> I thought so, but now see it's not yet. I'll try to get that to work and follow up here.

This is finally all done.

- The `tests.yml` workflow runs unit tests on any commit. 
- `docs.yml` tests whether the docs build when a review is requested on a PR that targets `master`.
- `docs.yml` additionally builds and deploys the docs to the `gh-pages` branch (which updates github.io) when a PR is merged into `master`. 
- The `README` shows whether the tests and the docs (build) are passing.

I also added a couple templates for submitting new issues.

**Review by rbooth200 — COMMENTED** [review]
This looks good. 

The only query I have is whether we need any restrictions on concurrency. Github will run multiple copies of the actions in parallel, which might cause issues if there are multiple threads trying to build / push the docs at the same time. This could happen if multiple pull requests are being merged and might leave the docs out of date. See [here](https://jldec.me/preventing-concurrent-github-actions)

**Comment by jeffjennings** [comment]
> This looks good.
> 
> The only query I have is whether we need any restrictions on concurrency. Github will run multiple copies of the actions in parallel, which might cause issues if there are multiple threads trying to build / push the docs at the same time. This could happen if multiple pull requests are being merged and might leave the docs out of date. See [here](https://jldec.me/preventing-concurrent-github-actions)

Ok I set the concurrency on the docs workflow; it could queue docs being built if multiple requests are submitted at once, but that's fine. It would be good for us not to be rapidly merging multiple PRs into master anyway - the delay to build/deploy the docs is ~2-3 minutes.

**Comment by rbooth200** [comment]
I agree that it was a "should not occur" anyway but now its a "cannot occur", which is better. I'm going to accept this, feel free to merge when you want

**Review by rbooth200 — APPROVED** [review]

**Pull request merged**